File: backend/src/main/java/com/luma/entity/TicketType.java
package com.luma.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "ticket_types")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class TicketType {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "event_id", nullable = false)
    private Event event;

    @Column(nullable = false)
    private String name;

    @Column(columnDefinition = "NVARCHAR(1000)")
    private String description;

    @Column(nullable = false, precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal price = BigDecimal.ZERO;

    @Column(nullable = false)
    private Integer quantity;

    @Column(nullable = false)
    @Builder.Default
    private Integer soldCount = 0;

    @Column(nullable = false)
    @Builder.Default
    private Integer maxPerOrder = 10;

    private LocalDateTime saleStartDate;

    private LocalDateTime saleEndDate;

    @Column(nullable = false)
    @Builder.Default
    private Boolean isVisible = true;

    @Column(nullable = false)
    @Builder.Default
    private Integer displayOrder = 0;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public boolean isFree() {
        return price == null || price.compareTo(BigDecimal.ZERO) == 0;
    }

    public boolean isAvailable() {
        return getAvailableQuantity() > 0;
    }

    public Integer getAvailableQuantity() {
        return quantity - soldCount;
    }

    public boolean isSaleActive() {
        LocalDateTime now = LocalDateTime.now();

        if (saleStartDate != null && now.isBefore(saleStartDate)) {
            return false;
        }

        if (saleEndDate != null && now.isAfter(saleEndDate)) {
            return false;
        }

        return true;
    }

    public boolean canPurchase(int requestedQuantity) {
        if (!isVisible) return false;
        if (!isSaleActive()) return false;
        if (requestedQuantity > maxPerOrder) return false;
        if (requestedQuantity > getAvailableQuantity()) return false;
        return true;
    }

    public String getStatus() {
        if (!isVisible) return "HIDDEN";
        if (!isSaleActive()) {
            LocalDateTime now = LocalDateTime.now();
            if (saleStartDate != null && now.isBefore(saleStartDate)) {
                return "NOT_STARTED";
            }
            return "ENDED";
        }
        if (getAvailableQuantity() <= 0) return "SOLD_OUT";
        return "AVAILABLE";
    }
}

File: mobile/lib/shared/models/ticket_type.dart
class TicketType {
  const TicketType({
    required this.id,
    required this.eventId,
    required this.name,
    this.description,
    required this.price,
    this.currency = 'USD',
    required this.quantity,
    required this.soldCount,
    required this.availableQuantity,
    required this.maxPerOrder,
    this.saleStartDate,
    this.saleEndDate,
    required this.isVisible,
    required this.displayOrder,
    required this.status,
    required this.isFree,
  });

  factory TicketType.fromJson(Map<String, dynamic> json) {
    return TicketType(
      id: json['id']?.toString() ?? '',
      eventId: json['eventId']?.toString() ?? '',
      name: json['name'] as String? ?? '',
      description: json['description'] as String?,
      price: (json['price'] as num?)?.toDouble() ?? 0.0,
      currency: json['currency'] as String? ?? 'USD',
      quantity: (json['quantity'] as num?)?.toInt() ?? 0,
      soldCount: (json['soldCount'] as num?)?.toInt() ?? 0,
      availableQuantity: (json['availableQuantity'] as num?)?.toInt() ?? 0,
      maxPerOrder: (json['maxPerOrder'] as num?)?.toInt() ?? 10,
      saleStartDate: json['saleStartDate'] != null
          ? DateTime.tryParse(json['saleStartDate'] as String)
          : null,
      saleEndDate: json['saleEndDate'] != null
          ? DateTime.tryParse(json['saleEndDate'] as String)
          : null,
      isVisible: json['isVisible'] as bool? ?? true,
      displayOrder: (json['displayOrder'] as num?)?.toInt() ?? 0,
      status: json['status'] as String? ?? 'AVAILABLE',
      isFree: json['isFree'] as bool? ?? ((json['price'] as num?)?.toDouble() ?? 0) == 0,
    );
  }

  final String id;
  final String eventId;
  final String name;
  final String? description;
  final double price;
  final String currency;
  final int quantity;
  final int soldCount;
  final int availableQuantity;
  final int maxPerOrder;
  final DateTime? saleStartDate;
  final DateTime? saleEndDate;
  final bool isVisible;
  final int displayOrder;
  final String status;
  final bool isFree;

  bool get isAvailable => isVisible && availableQuantity > 0;
  bool get isSoldOut => availableQuantity <= 0;
  bool get isNotStarted => false;
  bool get isEnded => false;

  int get maxAllowedPurchase {
    if (!isAvailable) return 0;
    return availableQuantity < maxPerOrder ? availableQuantity : maxPerOrder;
  }
}
